apps/ex_nvr/lib/ex_nvr/http.ex

```diff
@@ -0,0 +1,110 @@
+defmodule ExNVR.HTTP do
+  @moduledoc false
+
+  @spec get(binary(), Keyword.t()) ::
+          {:ok, Finch.Response.t()} | {:error, term()}
+  def get(url, opts \\ []) do
+    username = opts[:username] || ""
+    password = opts[:password] || ""
+
+    opts = opts ++ [method: :get, url: url]
+
+    if username == "" or password == "" do
+      do_call(:get, url)
+    else
+      http_headers = [{"Authorization", "Basic " <> Base.encode64(username <> ":" <> password)}]
+      result = do_call(:get, url, http_headers)
+
+      with {:ok, %{status: 401} = resp} <- result,
+           {:ok, digest_header} <- build_digest_auth_header(resp, opts) do
+        do_call(:get, url, [digest_header])
+      else
+        _other ->
+          result
+      end
+    end
+  end
+
+  @spec build_digest_auth_header(map(), Keyword.t()) :: {:ok, tuple()} | {:error, binary()}
+  def build_digest_auth_header(resp, opts) do
+    with {:ok, digest_opts} <- digest_auth_opts(resp, opts),
+         digest_header <- {"Authorization", encode_digest(digest_opts)} do
+      {:ok, digest_header}
+    end
+  end
+
+  defp digest_auth_opts(%{headers: headers}, opts) do
+    headers
+    |> Enum.map(fn {key, value} -> {String.downcase(key), value} end)
+    |> Map.new()
+    |> Map.fetch("www-authenticate")
+    |> case do
+      {:ok, "Digest " <> digest} ->
+        {:ok,
+         %{
+           nonce: match_pattern("nonce", digest),
+           realm: match_pattern("realm", digest),
+           qop: match_pattern("qop", digest),
+           username: opts[:username],
+           password: opts[:password],
+           method: Atom.to_string(opts[:method]) |> String.upcase(),
+           path: URI.parse(opts[:url]).path
+         }}
+
+      _other ->
+        {:error, "Digest auth is disabled or not supported"}
```

Comment:
It doesn't seem to me like an error if the Digest header is not available. Return empty string or just make this function build the whole digest header.